src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -703,11 +703,14 @@ public static IEnumerable<TSource> SkipWhile<TSource>(this IEnumerable<TSource>
 
         private static IEnumerable<TSource> SkipWhileIterator<TSource>(IEnumerable<TSource> source, Func<TSource, bool> predicate)
         {
-            bool yielding = false;
-            foreach (TSource element in source)
+            using (IEnumerator<TSource> e = source.GetEnumerator())
             {
-                if (!yielding && !predicate(element)) yielding = true;
-                if (yielding) yield return element;
+                do
+                    if (!e.MoveNext()) yield break;
+                while (predicate(e.Current));
+                do
+                    yield return e.Current;
+                while (e.MoveNext());
```

Comment:
Hmm, the old code was easier to read and maintain IMHO. What kind of speedup do you see here? I'd imagine it wouldn't be very significant vs. general overhead of iterators, but I'm prepared to be proven wrong.